README.md
# egg-session-mongo

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/egg-session-mongo.svg?style=flat-square
[npm-url]: https://npmjs.org/package/egg-session-mongo
[travis-image]: https://img.shields.io/travis/Mitscherlich/egg-session-mongo.svg?style=flat-square
[travis-url]: https://travis-ci.org/Mitscherlich/egg-session-mongo
[codecov-image]: https://img.shields.io/codecov/c/github/Mitscherlich/egg-session-mongo.svg?style=flat-square
[codecov-url]: https://codecov.io/github/Mitscherlich/egg-session-mongo?branch=master
[david-image]: https://img.shields.io/david/Mitscherlich/egg-session-mongo.svg?style=flat-square
[david-url]: https://david-dm.org/Mitscherlich/egg-session-mongo
[snyk-image]: https://snyk.io/test/npm/egg-session-mongo/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/egg-session-mongo
[download-image]: https://img.shields.io/npm/dm/egg-session-mongo.svg?style=flat-square
[download-url]: https://npmjs.org/package/egg-session-mongo

A session extension for store session in MongoDB.

see Chinese version: [中文](README.zh_CN.md)

## Install

```bash
$ npm i egg-session-mongo --save
```

## Usage

```js
// {app_root}/config/plugin.js
exports.sessionMongo = {
  enable: true,
  package: 'egg-session-mongo',
};
```

or you can overwrite you `MongoStore` options in `app.js`.

```js
// {app_root}/app.js
module.exports = app => {
  const MongoStore = require('egg-session-mongo')(app);

  app.beforeStart(async () => {
    app.sessionStore = new MongStore(/* options */);
  });
};
```

see [app.js](app.js) for more detail.

## Configuration

```js
// {app_root}/config/config.default.js
exports.sessionMongo = {
  url: 'mongodb://127.0.0.1/test'
};
```

see [config/config.example.js](config/config.example.js) for more detail.

## Questions & Suggestions

Please open an issue [here](https://github.com/eggjs/egg/issues).
or open an issue just at this repo's [issue](https://github.com/Mitscherlich/egg-session-mongo/issues) for more timely reply

## License

[MIT](LICENSE)


CHANGELOG.md
# Changelog for [egg-session-mongo](https://github.com/Mitscherlich/egg-session-mongo)

## 21 Mar 2018 (2.0.0)
# Rafctor session store core to [koa-session-mongo](https://github.com/Mitscherlich/koa-session-mongo). Now you can use it in Koa2.

## 10 Mar 2018 (1.0.1)
* Fisrt stable version.


test/session-mongo.test.js
'use strict';

const sleep = require('mz-modules/sleep');
const request = require('supertest');
const assert = require('assert');
const mm = require('egg-mock');

describe('test/session-mongo.test.js', () => {
  let app;
  let agent;

  before(() => {
    app = mm.app({ baseDir: 'apps/session-mongo-test' });
    return app.ready();
  });

  beforeEach(() => {
    agent = request.agent(app.callback());
  });

  after(() => app.close());
  afterEach(mm.restore);

  it('should get empty session and do not set cookie when session not populated', async () => {
    await agent
      .get('/get')
      .expect(200)
      .expect({})
      .expect(res => {
        assert(!res.header['set-cookie'].join('').match(/EGG_SESS/));
      });
  });

  it('should ctx.session= change the session', async () => {
    await agent
      .get('/set?foo=bar')
      .expect(200)
      .expect({ foo: 'bar' })
      .expect('set-cookie', /EGG_SESS=.*?;/);
  });

  it('should ctx.session.key= change the session', async () => {
    await agent
      .get('/set?key=foo&foo=bar')
      .expect(200)
      .expect({ key: 'foo', foo: 'bar' })
      .expect('set-cookie', /EGG_SESS=.*?;/);

    await agent
      .get('/setKey?key=bar')
      .expect(200)
      .expect({ key: 'bar', foo: 'bar' })
      .expect('set-cookie', /EGG_SESS=.*?;/);
  });

  it('should ctx.session=null remove the session', async () => {
    await agent
      .get('/set?key=foo&foo=bar')
      .expect(200)
      .expect({ key: 'foo', foo: 'bar' })
      .expect('set-cookie', /EGG_SESS=.*?;/);

    await agent
      .get('/remove')
      .expect(204)
      .expect('set-cookie', /EGG_SESS=;/);

    await agent
      .get('/get')
      .expect(200)
      .expect({});
  });

  it('should ctx.session.maxAge= change maxAge', async () => {
    await agent
      .get('/set?key=foo&foo=bar')
      .expect(200)
      .expect({ key: 'foo', foo: 'bar' })
      .expect('set-cookie', /EGG_SESS=.*?;/);

    let cookie;

    await agent
      .get('/maxAge?maxAge=100')
      .expect(200)
      .expect({ key: 'foo', foo: 'bar' })
      .expect(res => {
        cookie = res.headers['set-cookie'].join(';');
        assert(cookie.match(/EGG_SESS=.*?;/));
        assert(cookie.match(/expires=/));
      });

    await sleep(200);

    await agent
      .get('/get')
      .expect(200)
      .expect({});

    await request(app.callback())
      .get('/get')
      .set('cookie', cookie)
      .expect(200)
      .expect({});
  });
});


test/fixtures/apps/session-mongo-test/app/router.js
'use strict';

module.exports = app => {
  app.get('/get', 'home.get');
  app.get('/set', 'home.set');
  app.get('/setKey', 'home.setKey');
  app.get('/remove', 'home.remove');
  app.get('/maxAge', 'home.maxAge');
};


test/fixtures/apps/session-mongo-test/app/controller/home.js
'use strict';

exports.get = async ctx => { ctx.body = ctx.session; };

exports.set = async ctx => {
  ctx.session = ctx.query;
  ctx.body = ctx.session;
};

exports.setKey = async ctx => {
  ctx.session.key = ctx.query.key;
  ctx.body = ctx.session;
};

exports.remove = async ctx => {
  ctx.session = null;
  // delete ctx.session;
  ctx.body = ctx.session;
};

exports.maxAge = async ctx => {
  ctx.session.maxAge = Number(ctx.query.maxAge);
  ctx.body = ctx.session;
};


config/config.example.js
'use strict';

/**
 * egg-session-mongo default config
 * @member Config#sessionMongo
 * @property {string}         [url='mongodb://127.0.0.1/test']  - MongoDB server url
 * @property {string|Object}  [db='test']                       - `String` for a new connection or a exist mongoClient will be reuse if is an object
 * @property {string}         [collection='sessions']           - Session store collection name
 * @property {Object}         [mongooseConnection]              - Re-use existing or upcoming mongoose connection
 * @property {number}         [ttl=1209600]                     - Session expire time
 */
exports.sessionMongo = {
  url: 'mongodb://127.0.0.1',
  db: 'test',
  collection: 'sessions',
  mongooseConnection: null, // uncomment this to re-use existing or upcoming mongoose connection
  ttl: 14 * 24 * 60 * 60,   // = 14 days. Default
};


lib/session.js
'use strict';

const EventEmitter = require('events').EventEmitter;
const MongoClient = require('mongodb').MongoClient;
const Util = require('util');

/* istanbul ignore next */
function defaultSerializeFunction(session) {
  // Copy each property of the session to a new object
  const obj = {};
  let prop;

  for (prop in session) {
    if (prop === 'cookie') {
      // Convert the cookie instance to an object, if possible
      // This gets rid of the duplicate object under session.cookie.data property
      obj.cookie = session.cookie.toJSON ? session.cookie.toJSON() : session.cookie;
    } else {
      obj[prop] = session[prop];
    }
  }

  return obj;
}

/* istanbul ignore next */
function computeTransformFunctions(options, defaultStringify) {
  if (options.serialize || options.unserialize) {
    return {
      serialize: options.serialize || defaultSerializeFunction,
      unserialize: options.unserialize || (x => x),
    };
  }

  if (options.stringify === false || defaultStringify === false) {
    return {
      serialize: defaultSerializeFunction,
      unserialize: x => x,
    };
  }

  if (options.stringify === true || defaultStringify === true) {
    return {
      serialize: JSON.stringify,
      unserialize: JSON.parse,
    };
  }
}

module.exports = app => {
  class MongoStore extends EventEmitter {
    constructor(options) {
      super();

      /* Options */
      this.ttl = options.ttl || 1209600; // 14 days
      this.collectionName = options.collection || 'sessions';
      this.autoRemove = options.autoRemove || 'native';
      this.autoRemoveInterval = options.autoRemoveInterval || 10;
      this.transformFunctions = computeTransformFunctions(options, true);

      /* istanbul ignore if */
      if (!options.db) {
        throw new Error(`[egg-session-mongo] must provide 'db' in options but got '${options.db}'`);
      }

      if (typeof options.db === 'string') {
        this.dbName = options.db;
      }

      this.options = options;

      this.changeState('init');

      const newConnectionCallback = (err, client) => {
        /* istanbul ignore if */
        if (err) {
          this.connectionFailed(err);
        } else {
          this.handleNewConnectionAsync(client.db(this.dbName));
        }
      };

      /* istanbul ignore next */
      if (options.url) {
        // New native connection using url + mongoOptions
        MongoClient.connect(options.url, options.mongoOptions || {}, newConnectionCallback);
      } else if (options.mongooseConnection) {
        // Re-use existing or upcoming mongoose connection
        if (options.mongooseConnection.readyState === 1) {
          this.handleNewConnectionAsync(options.mongooseConnection.db);
        } else {
          options.mongooseConnection.once('open', () => this.handleNewConnectionAsync(options.mongooseConnection.db));
        }
      } else if (options.db && options.db.listCollections) {
        // Re-use existing or upcoming native connection
        if (options.db.openCalled || options.db.openCalled === undefined) { // OpenCalled is undefined in mongodb@2.x
          this.handleNewConnectionAsync(options.db);
        } else {
          options.db.open(newConnectionCallback);
        }
      } else if (options.dbPromise) {
        options.dbPromise
          .then(db => this.handleNewConnectionAsync(db))
          .catch(err => this.connectionFailed(err));
      } else {
        throw new Error('Connection strategy not found');
      }

      this.changeState('connecting');
    }

    /* istanbul ignore next */
    connectionFailed(err) {
      this.changeState('disconnected');
      throw err;
    }

    /* istanbul ignore next */
    handleNewConnectionAsync(db) {
      this.db = db;
      return this
        .setCollection(db.collection(this.collectionName))
        .setAutoRemoveAsync()
        .then(() => this.changeState('connected'));
    }

    /* istanbul ignore next */
    setAutoRemoveAsync() {
      const removeQuery = { expires: { $lt: new Date() } };
      switch (this.autoRemove) {
        case 'native':
          return this.collection.createIndex({ expires: 1 }, { expireAfterSeconds: 0 });
        case 'interval':
          this.timer = setInterval(() => this.collection.remove(removeQuery, { w: 0 }), this.autoRemoveInterval * 1000 * 60);
          this.timer.unref();
          return Promise.resolve();
        default:
          return Promise.resolve();
      }
    }

    changeState(newState) {
      if (newState !== this.state) {
        this.state = newState;
        this.emit(newState);
      }
    }

    setCollection(collection) {
      /* istanbul ignore of */
      if (this.timer) {
        clearInterval(this.timer);
      }

      this.collectionReadyPromise = undefined;
      this.collection = collection;

      return this;
    }

    /* istanbul ignore next */
    collectionReady() {
      let promise = this.collectionReadyPromise;
      if (!promise) {
        promise = new Promise((resolve, reject) => {
          if (this.state === 'connected') {
            return resolve(this.collection);
          }
          if (this.state === 'connecting') {
            return this.once('connected', () => resolve(this.collection));
          }
          reject(new Error('Not connected'));
        });
        this.collectionReadyPromise = promise;
      }
      return promise;
    }

    /* istanbul ignore next */
    computeStorageId(sessionId) {
      if (this.options.transformId && typeof this.options.transformId === 'function') {
        return this.options.transformId(sessionId);
      }
      return sessionId;
    }

    /* Public API */

    async get(sid) {
      return await this.collectionReady()
        .then(collection => collection.findOne({
          _id: this.computeStorageId(sid),
          $or: [
            { expires: { $exists: false } },
            { expires: { $gt: new Date() } },
          ],
        }))
        .then(session => {
          if (session) {
            const s = this.transformFunctions.unserialize(session.session);
            if (this.options.touchAfter > 0 && session.lastModified) {
              s.lastModified = session.lastModified;
            }
            this.emit('get', sid);
            return s;
          }
        });
    }

    async set(sid, session, maxAge) {
      // Removing the lastModified prop from the session object before update
      if (this.options.touchAfter > 0 && session && session.lastModified) {
        delete session.lastModified;
      }

      const s = { _id: this.computeStorageId(sid), session: this.transformFunctions.serialize(session) };

      if (session && session._expire) {
        s.expires = new Date(session._expire);
      } else {
        // If there's no expiration date specified, it is
        // browser-session cookie or there is no cookie at all,
        // as per the connect docs.
        //
        // So we set the expiration to two-weeks from now
        // - as is common practice in the industry (e.g Django) -
        // or the default specified in the options.
        s.expires = new Date(Date.now() + ((maxAge || this.ttl) * 1000));
      }

      if (this.options.touchAfter > 0) {
        s.lastModified = new Date();
      }

      return await this.collectionReady()
        .then(collection => collection.update({ _id: this.computeStorageId(sid) }, s, { upsert: true }))
        .then(rawResponse => {
          if (rawResponse.result) {
            rawResponse = rawResponse.result;
          }
          if (rawResponse && rawResponse.upserted) {
            this.emit('create', sid);
          } else {
            this.emit('update', sid);
          }
          this.emit('set', sid);
        });
    }

    /* istanbul ignore next */
    async touch(sid, session) {
      const updateFields = {};
      const touchAfter = this.options.touchAfter * 1000;
      const lastModified = session.lastModified ? session.lastModified.getTime() : 0;
      const currentDate = new Date();

      // If the given options has a touchAfter property, check if the
      // current timestamp - lastModified timestamp is bigger than
      // the specified, if it's not, don't touch the session
      if (touchAfter > 0 && lastModified > 0) {
        const timeElapsed = currentDate.getTime() - session.lastModified;

        if (timeElapsed < touchAfter) {
          return;
        }
        updateFields.lastModified = currentDate;
      }

      if (session && session.cookie && session.cookie.expires) {
        updateFields.expires = new Date(session.cookie.expires);
      } else {
        updateFields.expires = new Date(Date.now() + (this.ttl * 1000));
      }

      return await this.collectionReady()
        .then(collection => collection.update({ _id: this.computeStorageId(sid) }, { $set: updateFields }))
        .then(result => {
          if (result.nModified === 0) {
            throw new Error('Unable to find the session to touch');
          } else {
            this.emit('touch', sid, session);
          }
        });
    }

    /* istanbul ignore next */
    async destroy(sid) {
      return await this.collectionReady()
        .then(collection => collection.remove({ _id: this.computeStorageId(sid) }))
        .then(() => this.emit('destroy', sid));
    }

    /* istanbul ignore next */
    async length() {
      return await this.collectionReady()
        .then(collection => collection.count({}));
    }

    /* istanbul ignore next */
    async clear() {
      return await this.collectionReady()
        .then(collection => collection.drop());
    }

    /* istanbul ignore next */
    async close() {
      if (this.db) {
        this.db.close();
      }
    }
  }

  /* Deprecated APIs    */
  /* istanbul ignore if */
  if (app.config.sessionMongo.fallbackMemory && app.config.sessionMongo.fallbackMemory !== false) {
    Util.deprecate(() => console.log('[egg-session-mongo] option `fallbackMemory` is deprecated. Just disable this plugin in `plugin.default.js`.'));
  }

  return MongoStore;
};
